pkg/resolvers/k8s/streamer.go

```diff
@@ -33,79 +41,88 @@ func startWatchingEndpointsChanges(
-		innerCancel()
+		defer innerCancel()
+
+		if err := proxy(innerCtx, json.NewDecoder(stream), s.changeCh); err != nil {
+			s.errCh <- err
+		}
 	}()
 
-	return nil
+	return s, nil
 }
 
-type eventType string
-
-const (
-	added    eventType = "ADDED"
-	modified eventType = "MODIFIED"
-	deleted  eventType = "DELETED"
-	failed   eventType = "ERROR"
-)
+func (s *streamer) ResultChans() (<-chan change, <-chan error) {
+	return s.changeCh, s.errCh
+}
 
-// event represents a single event to a watched resource.
+// We don't want to use special, internal decoder, so we need to have all typed.
 type event struct {
-	Type   eventType `json:"type"`
-	Object endpoints `json:"object"`
+	Type   watch.EventType
+	Object eventObject
+}
+
+type eventObject struct {
+	// If type == Error.
+	*metav1.Status
+	// If type == Added, Modified or Deleted.
+	*v1.Endpoints
 }
 
-// proxyAllEvents gets events in loop and proxies to eventsCh. If event include some error it always returns, because
-// watchers.Next errors are meant to irrecoverable.
-func proxyAllEvents(ctx context.Context, decoder *json.Decoder, eventsCh chan<- watchResult) {
+// proxy receives events in loop and proxies to changeCh. If event include some error, or stream errors it always returns
+// error, because watchers.Next errors are meant to irrecoverable. On graceful EOF, or cancelled context, no error is returned.
+func proxy(ctx context.Context, decoder *json.Decoder, endpCh chan<- change) error {
+	var event event
+
 	for ctx.Err() == nil {
-		var eventErr error
-		var got event
 		// Blocking read.
-		if err := decoder.Decode(&got); err != nil {
+		if err := decoder.Decode(&event); err != nil {
 			if ctx.Err() != nil {
 				// Stopping state.
-				return
+				return nil
 			}
 			switch err {
 			case io.EOF:
-				// Watch closed normally - weird.
-				eventErr = errors.Wrap(err, "EOF during watch stream event decoding")
+				// Stream closed gracefully.
+				return io.EOF
 			case io.ErrUnexpectedEOF:
-				eventErr = errors.Wrap(err, "Unexpected EOF during watch stream event decoding")
+				return errors.Wrap(err, "unexpected EOF during watch stream event decoding")
 			default:
-				eventErr = errors.Wrap(err, "Unable to decode an event from the watch stream")
+				return errors.Wrap(err, "unable to decode an event from the watch stream")
 			}
 		}
 
-		if eventErr == nil {
-			switch got.Type {
-			case added, modified, deleted:
-			// All is fine.
-			case failed:
-				eventErr = errors.Errorf("%s: %s. Code: %d",
-					got.Object.Status,
-					got.Object.Message,
-					got.Object.Code,
+		switch event.Type {
+		case watch.Error:
+			if event.Object.Status != nil {
+				return errors.Errorf("%s: %s. Code: %d",
+					event.Object.Status.Status,
+					event.Object.Status.Message,
+					event.Object.Status.Code,
 				)
-			default:
-				eventErr = errors.Errorf("Got invalid watch event type: %v", got.Type)
 			}
+			return errors.Errorf("unexpected error object type %v", event.Object)
```

Comment:
Why aren't you using `errors.Wrapf` (here and below)? You use `errors.Wrap` in the code above.

Reply:
Am I wrapping any error here? (: